File: 14-TypeScript/Grey Multi-Tenant/packages/grey-dart/lib/src/config/grey_options.dart
/// Configuration options for the Grey SDK client.
library;

import 'package:grpc/grpc.dart';

/// Configuration options for connecting to the Grey API.
class GreyOptions {
  /// Creates new Grey client options.
  const GreyOptions({
    required this.host,
    this.port = 443,
    this.useTls = true,
    this.timeoutMs = 30000,
    this.channelOptions,
  });

  /// The host address of the Grey API server.
  final String host;

  /// The port to connect to. Defaults to 443.
  final int port;

  /// Whether to use TLS for the connection. Defaults to true.
  final bool useTls;

  /// Request timeout in milliseconds. Defaults to 30000 (30 seconds).
  final int timeoutMs;

  /// Optional custom channel options for gRPC.
  final ChannelOptions? channelOptions;

  /// Returns timeout as a Duration.
  Duration get timeout => Duration(milliseconds: timeoutMs);

  /// Creates channel credentials based on TLS setting.
  ChannelCredentials get credentials {
    return useTls
        ? const ChannelCredentials.secure()
        : const ChannelCredentials.insecure();
  }

  /// Creates a copy with optional modifications.
  GreyOptions copyWith({
    String? host,
    int? port,
    bool? useTls,
    int? timeoutMs,
    ChannelOptions? channelOptions,
  }) {
    return GreyOptions(
      host: host ?? this.host,
      port: port ?? this.port,
      useTls: useTls ?? this.useTls,
      timeoutMs: timeoutMs ?? this.timeoutMs,
      channelOptions: channelOptions ?? this.channelOptions,
    );
  }

  @override
  String toString() {
    return 'GreyOptions(host: $host, port: $port, useTls: $useTls, timeoutMs: $timeoutMs)';
  }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-scala/src/main/scala/com/grey/sdk/config/GreyOptions.scala
package com.grey.sdk.config

import io.grpc.{ManagedChannel, ManagedChannelBuilder}
import io.grpc.netty.shaded.io.grpc.netty.{GrpcSslContexts, NettyChannelBuilder}
import java.util.concurrent.TimeUnit
import scala.concurrent.duration.Duration

/** Configuration options for Grey SDK client.
  *
  * @param host
  *   The host address of the Grey API server
  * @param port
  *   The port to connect to (default: 443)
  * @param useTls
  *   Whether to use TLS for the connection (default: true)
  * @param timeoutMs
  *   Request timeout in milliseconds (default: 30000)
  */
final case class GreyOptions(
    host: String,
    port: Int = 443,
    useTls: Boolean = true,
    timeoutMs: Long = 30000
):
  require(host.nonEmpty, "Host must not be empty")
  require(port > 0 && port <= 65535, "Port must be between 1 and 65535")
  require(timeoutMs > 0, "Timeout must be positive")

  /** Returns timeout as a Duration. */
  def timeout: Duration = Duration(timeoutMs, TimeUnit.MILLISECONDS)

  /** Creates a copy with modified settings. */
  def withHost(newHost: String): GreyOptions = copy(host = newHost)
  def withPort(newPort: Int): GreyOptions = copy(port = newPort)
  def withTls(enabled: Boolean): GreyOptions = copy(useTls = enabled)
  def withTimeout(ms: Long): GreyOptions = copy(timeoutMs = ms)

object GreyOptions:
  /** Creates default options for local development. */
  def local(port: Int = 50051): GreyOptions =
    GreyOptions(host = "localhost", port = port, useTls = false)

  /** Creates default options for production. */
  def production(host: String): GreyOptions =
    GreyOptions(host = host, port = 443, useTls = true)
